File: competitive-programming/src/edu/algos/li21_dp/medium/Stocks_L122_BestTimeToBuyAndSellStockII.java
package edu.algos.li21_dp.medium;

public class Stocks_L122_BestTimeToBuyAndSellStockII {

	public static void main(String[] args) {
		Stocks_L122_BestTimeToBuyAndSellStockII obj = new Stocks_L122_BestTimeToBuyAndSellStockII();

		int[] prices = new int[] {7,1,5,3,6,4};
		System.out.println(obj.maxProfit(prices));

		System.out.println(obj.maxProfit(prices, 0, 1));
	}

	/*
	 * Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
	 * Unlimited transaction but one at a time.
	 * Idea: As we have to remember whether we bought previously or not we need to pass extra parameter to recursion.
	 * If we are buying it will contribute -ve to profit as profit = sell - buy. Hence add -ve price if buying.
	 * i/j 	7 1 5 3 6 4
	 * buy 	0 0
	 * sell 0
	 */
	public int maxProfit(int[] prices) {
		int n = prices.length;

		int[][] profits = new int[n+1][2];
		for (int i = n; i >= 0; i--) {
			for (int isBuy = 0; isBuy <= 1; isBuy++) {
				if(i == n){
					profits[i][isBuy] = 0; 	// out of market profit is 0
				}else if (isBuy == 1) { 	// buy
					int buy = -prices[i] + profits[i + 1][0];
					int notBuy = 0 + profits[i + 1][1];
					profits[i][isBuy] = Math.max(buy, notBuy);
				} else {				// sell
					int sell = prices[i] + profits[i + 1][1];
					int notSell = 0 + profits[i + 1][0];
					profits[i][isBuy] = Math.max(sell, notSell);
				}
			}
		}
		return profits[0][1];
	}

	public int maxProfit(int[] prices, int i, int isBuy) {
		int n = prices.length;
		if (i == n)
			return 0; // selling once market close gives 0 price
		if (isBuy == 1) {
			int buy = -prices[i] + maxProfit(prices, i + 1, 0);
			int notBuy = 0 + maxProfit(prices, i + 1, 1);
			return Math.max(buy, notBuy);
		} else {
			int sell = prices[i] + maxProfit(prices, i + 1, 1);
			int notSell = 0 + maxProfit(prices, i + 1, 0);
			return Math.max(sell, notSell);
		}
	}

}

File: dsa/concepts/15_dp/python/stocks_l122_best_time_to_buy_and_sell_stock_ii.py
# DP / Stock
#
# Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
# Unlimited transactions but one at a time
# Idea:
# We track whether we are allowed to buy (isBuy=1) or sell (isBuy=0)
# Buying contributes -price, selling contributes +price
# DP state: dp[i][isBuy] = max profit starting at index i with buy/sell state

def maxProfit(prices):
    n = len(prices)
    dp = [[0] * 2 for _ in range(n + 1)]

    for i in range(n - 1, -1, -1):
        for isBuy in range(2):
            if isBuy == 1:  # allowed to buy
                buy = -prices[i] + dp[i + 1][0]
                notBuy = dp[i + 1][1]
                dp[i][isBuy] = max(buy, notBuy)
            else:           # allowed to sell
                sell = prices[i] + dp[i + 1][1]
                notSell = dp[i + 1][0]
                dp[i][isBuy] = max(sell, notSell)

    return dp[0][1]


def maxProfitRecursive(prices, ind=0, isBuy=1):
    n = len(prices)
    if ind == n:
        return 0

    if isBuy == 1:
        buy = -prices[ind] + maxProfitRecursive(prices, ind + 1, 0)
        notBuy = maxProfitRecursive(prices, ind + 1, 1)
        return max(buy, notBuy)
    else:
        sell = prices[ind] + maxProfitRecursive(prices, ind + 1, 1)
        notSell = maxProfitRecursive(prices, ind + 1, 0)
        return max(sell, notSell)


# -------- Test --------
prices = [7, 1, 5, 3, 6, 4]

print(maxProfit(prices))                  # Expected: 7
print(maxProfitRecursive(prices, 0, 1))  # Expected: 7
